darts/models/nbeats.py

```diff
@@ -0,0 +1,410 @@
+    SEASONALITY = 3
+
+
+GTypes = NewType('GTypes', GType)
+
+
+class _TrendGenerator(nn.Module):
+
+    def __init__(self,
+                 expansion_coefficient_dim,
+                 output_dim):
+        super(_TrendGenerator, self).__init__()
+        self.T = torch.stack([(torch.arange(output_dim) / output_dim)**i for i in range(expansion_coefficient_dim)], 1)
+
+    def forward(self, x):
+        return torch.matmul(x, self.T.float().T)
+
+
+class _SeasonalityGenerator(nn.Module):
+
+    def __init__(self,
+                 output_dim):
+        super(_SeasonalityGenerator, self).__init__()
+        half_minus_one = int(output_dim / 2 - 1)
+        cos_vectors = [torch.cos(torch.arange(output_dim) * 2 * np.pi * i) for i in range(1, half_minus_one + 1)]
+        sin_vectors = [torch.sin(torch.arange(output_dim) * 2 * np.pi * i) for i in range(1, half_minus_one + 1)]
+        self.S = torch.stack([torch.ones(output_dim)] + cos_vectors + sin_vectors, 1)
+
+    def forward(self, x):
+        return torch.matmul(x, self.S.float().T)
+
+
+class _Block(nn.Module):
+
+    def __init__(self,
+                 num_layers: int,
+                 layer_width: int,
+                 expansion_coefficient_dim: int,
+                 input_length: int,
+                 output_length: int,
+                 g_type: GTypes):
+        """ PyTorch module implementing the basic building block of the N-BEATS architecture.
+
+        Parameters
+        ----------
+        num_layers
+            The number of fully connected layers preceding the final forking layers.
+        layer_width
+            The number of neurons that make up each fully connected layer.
+        expansion_coefficient_dim
+            The dimensionality of the waveform generator parameters, also known as expansion coefficients.
+            Only used if `generic_architecture` is set to `True`.
+        input_length
+            The length of the input sequence fed to the model.
+        output_length
+            The length of the forecast of the model.
+        g_type
+            The type of function that is implemented by the waveform generator.
+
+        Inputs
+        ------
+        x of shape `(batch_size, input_length)`
+            Tensor containing the input sequence.
+
+        Outputs
+        -------
+        x_hat of shape `(batch_size, input_length)`
+            Tensor containing the 'backcast' of the block, which represents an approximation of `x`
+            given the constraints of the functional space determined by `g`.
+        y_hat of shape `(batch_size, output_length)`
+            Tensor containing the forward forecast of the block.
+
+        """
+        super(_Block, self).__init__()
+
+        self.num_layers = num_layers
+        self.layer_width = layer_width
+        self.g_type = g_type
+        self.relu = nn.ReLU()
+
+        # fully connected stack before fork
+        self.linear_layer_stack_list = [nn.Linear(input_length, layer_width)]
```

Comment:
(random comment, not for this PR)
It would be nice to optionally support multihead attention with positional encoding here, instead of fully-connected.